File: bench/micro/fn_call.py
"""Python equivalent of fn_call.ap.
Free function called in a tight loop. Python's function-call
machinery is famously expensive — frame allocation per call.
"""

import time


def noop(x):
    return x


iters = 10_000_000
t0 = time.monotonic_ns()
acc = 0
for i in range(iters):
    acc = noop(i)
elapsed = time.monotonic_ns() - t0
print(f"iters={iters}")
print(f"acc={acc}")
print(f"elapsed_ns={elapsed}")

File: bench/micro/fn_call.go
// Go equivalent of fn_call.ap.
// Free-fn call overhead.
package main

import (
	"fmt"
	"time"
)

//go:noinline
func noop(x int) int {
	return x
}

func main() {
	iters := 10000000
	t0 := time.Now()
	acc := 0
	for i := 0; i < iters; i++ {
		acc = noop(i)
	}
	elapsed := time.Since(t0).Nanoseconds()
	fmt.Printf("iters=%d\n", iters)
	fmt.Printf("acc=%d\n", acc)
	fmt.Printf("elapsed_ns=%d\n", elapsed)
}
